File: java-spring/architecture-shell-cqrs/src/main/java/com/architecture/shell/cqrs/PipelineBehavior.java
package com.architecture.shell.cqrs;

/**
 * Interface for pipeline behaviors that wrap handler execution.
 * Behaviors implement cross-cutting concerns like validation, authorization, transactions, telemetry, caching.
 *
 * Requirements:
 * - FR-004: Pipeline behaviors execute in configured order
 * - FR-007: Behaviors apply selectively via BehaviorMatcher type guards
 * - BR-004: Recommended order: Validation → Authorization → Transaction → Telemetry → Resilience
 *
 * Execution Pattern:
 * 1. Behavior pre-processing (validation, transaction begin, start timer)
 * 2. Call next() to continue pipeline
 * 3. Behavior post-processing (transaction commit, log duration)
 *
 * Behavior Order:
 * - Configured via BehaviorRegistry.register(behavior, order)
 * - Lower order numbers execute first (outer decorators)
 * - Non-recommended orders log warning per BR-004
 *
 * Example:
 * <pre>
 * public class ValidationBehavior<TRequest, TResponse> implements PipelineBehavior<TRequest, TResponse> {
 *     private final IValidator<TRequest> validator;
 *
 *     public TResponse handle(TRequest request, RequestHandlerDelegate<TResponse> next) {
 *         var validationResult = validator.validate(request);
 *         if (validationResult.isFailure()) {
 *             throw new ValidationException(validationResult.errors());
 *         }
 *         return next.handle();
 *     }
 * }
 * </pre>
 *
 * Short-Circuiting:
 * - Behaviors may throw exceptions to abort pipeline
 * - Behaviors may return early without calling next() (e.g., cache hit)
 *
 * DDD Layer: Application Layer Infrastructure
 *
 * @param <TRequest>  The request type
 * @param <TResponse> The response type
 * @see BehaviorMatcher
 * @see UnitOfWorkBehavior
 * @see ValidationBehavior
 */
@FunctionalInterface
public interface PipelineBehavior<TRequest extends BaseRequest, TResponse> {

    /**
     * Handles the request by executing behavior logic and optionally calling the next delegate.
     *
     * @param request The request being processed
     * @param next    The delegate for the next behavior or handler in the pipeline
     * @return The response from this behavior or downstream pipeline
     * @throws RuntimeException to abort pipeline execution
     */
    TResponse handle(TRequest request, RequestHandlerDelegate<TResponse> next);
}

File: golang/architecture-shell-cqrs/pipeline_behavior.go
package cqrs

import "context"

// RequestHandlerDelegate represents the next step in the pipeline (next behavior or handler).
// Returns (response, nil) on success, (zero, error) on failure.
type RequestHandlerDelegate[TResponse any] func() (TResponse, error)

// PipelineBehavior represents a pipeline behavior for cross-cutting concerns.
// Wraps handler execution with pre/post logic.
//
// Common Behaviors:
//   - ValidationBehavior: Validates request payload; returns error on validation failure (short-circuits pipeline)
//   - AuthorizationBehavior: Checks permissions; returns error if denied
//   - UnitOfWorkBehavior: Opens transaction for commands; detects active transaction for nested commands; commits on success, rolls back on error
//   - CachingBehavior: Checks cache for queries; returns cached value on hit, executes handler on miss
//   - TelemetryBehavior: Logs request type, duration, status, errors
//   - ResilienceBehavior: Wraps handler with retry/timeout/circuit breaker logic
//
// Constraints:
//   - Behaviors execute in configured order (default: Validation → Authorization → Transaction → Telemetry → Resilience)
//   - Behaviors MAY short-circuit pipeline by returning error or returning early
//   - Behaviors MUST call next() to continue pipeline unless explicitly short-circuiting
type PipelineBehavior[TRequest BaseRequest, TResponse any] interface {
	// Handle executes cross-cutting logic before/after calling next.
	// ctx: Context for cancellation and deadline control
	// request: The request being processed
	// next: The next behavior or final handler in the pipeline
	// Returns: (response, nil) on success, (zero, error) on failure
	Handle(
		ctx context.Context,
		request TRequest,
		next RequestHandlerDelegate[TResponse],
	) (TResponse, error)

	// Order returns the execution order (lower values execute first).
	// Default: 100
	// Recommended: Validation=10, Authorization=20, Transaction=30, Telemetry=40, Caching=50, Resilience=60
	Order() int
}
